code/obj/machinery/computer/security/riot.dm

```diff
@@ -135,135 +140,199 @@
-				O.show_message("<span class='subtle'><span class='game say'><span class='name'>[src]</span> beeps, \"[src.auth_need - src.authorized.len] authorizations needed until Armory is opened.\"</span></span>", 2)
+				O.show_message("<span class='subtle'><span class='game say'><span class='name'>[src]</span> beeps, \"[src.auth_need - src.authorisations.len] authorizations needed until Armory is opened.\"</span></span>", 2)
+
+	proc/checkRequirements(var/mob/user, var/obj/item/W)
+		// Reset the variables.
+		hasID = FALSE
+		hasAccess = FALSE
+		mismatchedAppearance = FALSE
+		mismatchedID = FALSE
+		maxSecAccess = FALSE
+		existingAuthorisation = FALSE
+		nameOnFile = FALSE
+		printOnFile = FALSE
+
+		if (W == null) // No ID.
+			return
+
+		if (!istype(W, /obj/item/card/id)) // No ID.
+			return
+		else
+			hasID = TRUE
+
+		var/obj/item/card/id/ID = W
+
+		if (!ID:access) // No access.
+			return
+
+		var/list/cardaccess = ID:access
+		if (!istype(cardaccess, /list) || !length(cardaccess)) // No access.
+			return
+
+		if (access_securitylockers in ID:access) // Has Security Equipment access.
+			hasAccess = TRUE
+
+		if (ishuman(user))
+			var/mob/living/carbon/human/H = user
+			if (H.name != strip_html_tags(H.get_heard_name())) // User's appearance must match their voice.
+				mismatchedAppearance = TRUE
+				boutput(user, "[H.name] | [strip_html_tags(H.get_heard_name())]")
+				return
+
+		if (user.name != ID.registered) // Player name must match their ID's registered name.
+			mismatchedID = TRUE
+			return
+
+		if (access_maxsec in ID:access) // Are they the HoS, or do they have equivalent access?
+			maxSecAccess = TRUE
 
+		if (user.name in authorisations) // Check for authorisations that use the player's name.
+			nameOnFile = TRUE
+
+		for (var/x in authorisations)
+			var/list/params = authorisations[x]
+			if (user.bioHolder.fingerprints == params["prints"]) // Check for authorisations that use the player's fingerprints.
+				printOnFile = TRUE
+				break
+
+		// If both their name and fingerprint are on file, then they may repeal their authorisation.
+		// If either only their name or fingerprint are on file, and not the other, then the computer will deny them access.
+		if (nameOnFile == TRUE && printOnFile == TRUE)
+			existingAuthorisation = TRUE
 
 /obj/machinery/computer/riotgear/attack_hand(mob/user)
 	if (ishuman(user))
 		return src.Attackby(user:wear_id, user)
 	..()
 
-//kinda copy paste from shuttle auth :)
 /obj/machinery/computer/riotgear/attackby(var/obj/item/W, var/mob/user)
+
 	interact_particle(user,src)
+	src.add_fingerprint(user)
+
 	if(status & (BROKEN|NOPOWER))
 		return
 	if (!user)
 		return
 
 	if (istype(W, /obj/item/device/pda2) && W:ID_card)
 		W = W:ID_card
-	if (!istype(W, /obj/item/card/id))
-		boutput(user, "No ID given.")
-		return
 
-	if (!W:access) //no access
-		src.add_fingerprint(user)
-		boutput(user, "The access level of [W] is not high enough.")
+	checkRequirements(user, W)
```

Comment:
This seems like it could go wrong in a number of ways, for example if one person opens the panel without the proper credentials and then an HoS opens it won't the first person be be able to immediately deauth since the HoS's authorizations will be stored here?

Reply:
Any time any interaction is made with the computer's UI `checkRequirements()` is called, see line `293`, so anyone who doesn't have access (or fails any of the other checks) shouldn't be able to do that.

Reply:
Okay so why store it then? If the only reason is to cache the UI state then that's just going to cause the UI to display the wrong user's version if someone else clicks on it (remember that tgui calls `ui_data` periodically while the UI is open).

Reply:
You're entirely right, yes... I'm not quite sure how to prevent that; what would you recommend for storing that data per player?

Reply:
Honestly I probably wouldn't store it, it should be light enough to just call every time its needed (which it mostly is anyway)
If you really want to store it then you could do an associative list of mob refs to auth states and only update it on certain conditions, but that's probably overkill.